### ISSUE TITLE
Infinite suspense loop in useSuspenseQuery if the same query document is first skipped, then fetched in the same suspense boundary 

### ISSUE THREAD

**Issue by destin-estrela** [apollographql/apollo-client#11490]
### Issue Description

This usage of useSuspenseQuery will cause an infinite suspense loop. 
```tsx
useSuspenseQuery(ALL_PEOPLE, skipToken);
useSuspenseQuery(ALL_PEOPLE);
```
This is what I found when viewing the implementation.
1. Whenever you pass a `skipToken` to `useSuspenseQuery`, it changes the query options fetch policy to "standby".
2. When you call the second query, it changes the query options back to "cache-first". 
3. When the options changes from "standby"  to "cache-first", it initiates the promise to loading state and suspends.
```ts
// apollo-client/src/react/cache/QueryReference.ts:182
if (
	currentFetchPolicy === "standby" &&
	currentFetchPolicy !== watchQueryOptions.fetchPolicy // this is "cache-first"
) {
	this.initiateFetch(this.observable.reobserve(watchQueryOptions)); // this creates a new promise
```

4. The component unsuspends, thus remounting the component and calling the first query with the skip token again. 
5. It changes the fetch policy to "standby" and returns the cached result successfully
6. It calls the second query. It changes the options from "standby" to "cache-first", initiating the promise to loading state and suspends.......

Loop to infinity

### Link to Reproduction
Reproduction in code sandbox [here](https://codesandbox.io/p/devbox/black-hooks-2288km?layout=%257B%2522sidebarPanel%2522%253A%2522EXPLORER%2522%252C%2522rootPanelGroup%2522%253A%257B%2522direction%2522%253A%2522horizontal%2522%252C%2522contentType%2522%253A%2522UNKNOWN%2522%252C%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522id%2522%253A%2522ROOT_LAYOUT%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522UNKNOWN%2522%252C%2522direction%2522%253A%2522vertical%2522%252C%2522id%2522%253A%2522clrg3kp5200072v6di8tct9kv%2522%252C%2522sizes%2522%253A%255B100%252C0%255D%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522EDITOR%2522%252C%2522direction%2522%253A%2522horizontal%2522%252C%2522id%2522%253A%2522EDITOR%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522EDITOR%2522%252C%2522id%2522%253A%2522clrg3kp5100022v6d68wloqsz%2522%257D%255D%257D%252C%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522SHELLS%2522%252C%2522direction%2522%253A%2522horizontal%2522%252C%2522id%2522%253A%2522SHELLS%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522SHELLS%2522%252C%2522id%2522%253A%2522clrg3kp5100042v6dj5c2gk1o%2522%257D%255D%252C%2522sizes%2522%253A%255B100%255D%257D%255D%257D%252C%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522DEVTOOLS%2522%252C%2522direction%2522%253A%2522vertical%2522%252C%2522id%2522%253A%2522DEVTOOLS%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522DEVTOOLS%2522%252C%2522id%2522%253A%2522clrg3kp5100062v6d7m1jq4cn%2522%257D%255D%252C%2522sizes%2522%253A%255B100%255D%257D%255D%252C%2522sizes%2522%253A%255B50%252C50%255D%257D%252C%2522tabbedPanels%2522%253A%257B%2522clrg3kp5100022v6d68wloqsz%2522%253A%257B%2522tabs%2522%253A%255B%257B%2522id%2522%253A%2522clrg3kp5100012v6dts10rrnk%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522FILE%2522%252C%2522filepath%2522%253A%2522%252FREADME.md%2522%252C%2522state%2522%253A%2522IDLE%2522%257D%252C%257B%2522id%2522%253A%2522clrg3n4w5005c2v6dw01c2jsu%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522FILE%2522%252C%2522initialSelections%2522%253A%255B%257B%2522startLineNumber%2522%253A57%252C%2522startColumn%2522%253A1%252C%2522endLineNumber%2522%253A84%252C%2522endColumn%2522%253A9%257D%255D%252C%2522filepath%2522%253A%2522%252Fsrc%252Findex.jsx%2522%252C%2522state%2522%253A%2522IDLE%2522%257D%255D%252C%2522id%2522%253A%2522clrg3kp5100022v6d68wloqsz%2522%252C%2522activeTabId%2522%253A%2522clrg3n4w5005c2v6dw01c2jsu%2522%257D%252C%2522clrg3kp5100062v6d7m1jq4cn%2522%253A%257B%2522tabs%2522%253A%255B%257B%2522id%2522%253A%2522clrg3kp5100052v6dp1cny2zi%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522TASK_PORT%2522%252C%2522taskId%2522%253A%2522start%2522%252C%2522port%2522%253A3000%252C%2522path%2522%253A%2522%252F%2522%257D%255D%252C%2522id%2522%253A%2522clrg3kp5100062v6d7m1jq4cn%2522%252C%2522activeTabId%2522%253A%2522clrg3kp5100052v6dp1cny2zi%2522%257D%252C%2522clrg3kp5100042v6dj5c2gk1o%2522%253A%257B%2522tabs%2522%253A%255B%257B%2522id%2522%253A%2522clrg3kp5100032v6dw7zslghi%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522TASK_LOG%2522%252C%2522taskId%2522%253A%2522start%2522%257D%255D%252C%2522id%2522%253A%2522clrg3kp5100042v6dj5c2gk1o%2522%252C%2522activeTabId%2522%253A%2522clrg3kp5100032v6dw7zslghi%2522%257D%257D%252C%2522showDevtools%2522%253Atrue%252C%2522showShells%2522%253Afalse%252C%2522showSidebar%2522%253Atrue%252C%2522sidebarPanelSize%2522%253A15%257D)

### Reproduction Steps

Easy, just call the same query document using `useSuspenseQuery` twice in a row, one with skip token, one without.

```tsx
useSuspenseQuery(ALL_PEOPLE, skipToken);
useSuspenseQuery(ALL_PEOPLE);
```

Here is a unit test to recreate it that you can run in the apollo client repo.
```tsx
  it.only("suspends a query that was skipped once and returns results", async () => {
    const { query, mocks } = useSimpleQueryCase();

    const Component = () => {
      useSuspenseQuery(query, skipToken);
      const result = useSuspenseQuery(query);
      ProfiledApp.replaceSnapshot(result);
      return <div>{result.data.greeting}</div>;
    };

    const App = () => {
      return (
        <Suspense fallback={<div>loading</div>}>
          <ErrorBoundary fallback={<div>Error</div>}>
            <Component />
          </ErrorBoundary>
        </Suspense>
      );
    };

    const ProfiledApp = profile<
      UseSuspenseQueryResult<SimpleQueryData, OperationVariables>
    >({
      Component: App,
      snapshotDOM: true,
    });

    const client = new ApolloClient({
      cache: new InMemoryCache(),
      link: new MockLink(mocks),
    });

    render(<ProfiledApp />, {
      wrapper: ({ children }) => (
        <ApolloProvider client={client}>{children}</ApolloProvider>
      ),
    });

    
    {
      // ensure the hook suspends immediately
      const { withinDOM, snapshot } = await ProfiledApp.takeRender();
      expect(withinDOM().getByText("loading")).toBeInTheDocument();
      expect(snapshot).toBeUndefined();
    }
    {
      const { withinDOM, snapshot } = await ProfiledApp.takeRender();
      expect(withinDOM().queryByText("loading")).not.toBeInTheDocument();
      expect(withinDOM().getByText("Hello")).toBeInTheDocument();
      expect(snapshot).toMatchObject({
        ...mocks[0].result,
        networkStatus: NetworkStatus.ready,
        error: undefined,
      });
    }
  });
```

**Comment by jerelmiller** [comment]
Hey @destin-estrela 👋 

tl;dr; this is expected behavior if you've got 2 suspense hooks with the same query/variables combination. If you need to differentiate them so they don't end up clobbering each other (like you're seeing here), pass a `queryKey` option to one of those queries and this will fix your issue. See [this section](https://www.apollographql.com/docs/react/data/suspense#distinguishing-between-queries-with-querykey) in our Suspense docs about differentiating queries.

I've updated your [reproduction](https://codesandbox.io/p/devbox/epic-hofstadter-mxdj3n) by passing a `queryKey` option to one of them and the issue goes away.

---

As for the longer explanation, Suspense works by `throw`ing a `Promise` during render. React uses this promise to render your nearest Suspense boundary, then try rendering again once all caught promsies have been resolved.

Because we throw during render, we have to cache that promise somewhere so that we don't end up creating new promises each time the component is re-rendered, otherwise your component will suspend indefinitely. Unfortunately if you suspend on first render, React doesn't store state between renders via `useRef` or `useState`, so we need a way to identify that promise in our suspense cache. To do this, we use a combination of the `client`, `query`, and `variables` as a cache key to try and uniquely identify that promise in the suspense cache, that way when your component is done suspending, we get a reference back to the same promise. (Our actual implementation is a bit more than just a promise. We store it in a "query reference" instance, but for this issue, I'll keep the mental model simple.)

As you can see in your example, you're using the same query/variables combination in both hooks, therefore each `useSuspenseQuery` hook thinks its working with the same underlying reference. `queryKey` was designed for this scenario so that you can add additional information to the cache key to uniquely identify it from another hook with the same query/variables. This way each hook will have its own underlying reference rather than trying to share the same.

---

Just out of total curiosity, do you actually put 2 `useSuspenseQuery` hooks right next to each other like this with one skipped, or was this a means to show a reproduction? We wouldn't recommend this in a production application which is why I ask 🙂.

Thanks for the question!

**Comment by jerelmiller** [comment]
I'm going to go ahead and close this issue as working as designed 🙂

**Comment by destin-estrela** [comment]
@jerelmiller Thank you! Looks like a glossed over one very critical section of the documentation. To answer your question, no, I don't put two suspense hooks like this next to each other! They happened quite far away from each other in my actual code base.

**Comment by jerelmiller** [comment]
@destin-estrela ok awesome! I figured that was probably the case and you were using this as a means for reproduction, so just thought I'd check! Hope the `queryKey` works well for you!
